### PR TITLE
First implementation pull request.

### PR DESCRIPTION
**Pull request by Yatskov**
@mxsnq please review current changes.

### CODE REVIEW

File: routes/click.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const express = require('express');
+const utils = require('../helpers/utils');
+const router = express.Router();
+
+
+async function action(page, request) {
+    return Promise.all([
```
Comment (mxsnq) [inline routes/click.js]: - I think you should await for promises and form regular response
- we should consider waiting for navigation

File: routes/screenshot.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        screenshot: screenshot
+    };
+}
+
+// Method that returns screenshots of pages
+// more description of options you can see on github:
+// https://github.com/GoogleChrome/puppeteer/blob/v1.19.0/docs/api.md#pagescreenshotoptions
+router.post('/', async function (req, res, next) {
+
+    if (!(typeof action === "function" && action.length >= 1)) {
```
Comment (mxsnq) [inline routes/screenshot.js]: Do you still need this?

File: helpers/utils.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+/***
+ * This function returns a page from browser context or create new page or even context if page_id or context_id are
+ * none. If no context or now page found throw an error.
+ * @param browser
+ * @param context_id - identifier of context to find.
+ * @param page_id - identifier of page to find.
+ * @returns {Promise<Page>}
+ */
+exports.getBrowserPage = async function getBrowserPage(browser, context_id, page_id) {
```
Comment (mxsnq) [inline helpers/utils.js]: I find confusing having both `getPage` and `getBrowserPage`. 
I'd suggest renaming `getPage` to `findPageInContext`, what do you think?

File: helpers/utils.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        return await context.newPage();
+    } else {
+        let context = await browser.createIncognitoBrowserContext();
+        return await context.newPage();
+    }
+};
+
+exports.perfomAction = async function perfomAction(request, action) {
+    let lock = request.app.get('lock');
+    let page = await exports.getBrowserPage(request.app.get('browser'), request.query.context_id, request.query.page_id);
```
Comment (mxsnq) [inline helpers/utils.js]: I'd suggest defining all functions first, then define all exports.
Don't know what are good practices here though.

File: routes/click.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        page.waitFor(request.body.waitOptions.selectorOrTimeout),
```
Comment (mxsnq) [inline routes/click.js]: Provide default value for selectorOrTimeout

File: routes/scroll.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+//  body = {
+//     "selector": "", //<string> A selector to search for element to scroll
+//     "waitOptions": {
+//         // if selectorOrFunctionOrTimeout is a string, then the first argument is treated as a selector or xpath, depending on whether or not it starts with '//', and the method is a shortcut for page.waitForSelector or page.waitForXPath
+//         // if selectorOrFunctionOrTimeout is a number, then the first argument is treated as a timeout in milliseconds and the method returns a promise which resolves after the timeout
+//         "selectorOrTimeout":...,
+//     }
+//  }
+router.post('/', async function (req, res, next) {
+    try {
```
Comment (mxsnq) [inline routes/scroll.js]: Add bad request response for missing selector

File: routes/goto.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+const express = require('express');
+const utils = require('../helpers/utils');
+const router = express.Router();
+
+
+async function action(page, request) {
+    await page.goto(request.body.url);
```
Comment (mxsnq) [inline routes/goto.js]: - add options in goto
- consider adding similar goForward and goBack actions

**Comment by mxsnq** [comment]
Also: add endpoint to close browser contexts.

File: routes/click.js (hunk 1 above)
Comment (Yatskov) [reply routes/click.js]: The pattern for doing this was brought from [pupeteer documentation](https://github.com/GoogleChrome/puppeteer/blob/v1.19.0/docs/api.md#pageclickselector-options).

Comment (mxsnq) [reply routes/click.js]: The pattern you are referring to uses waitForNavigation. You are using waitFor. Their promises resolve to different objects.
Nonetheless, I think this action should capture page state after click regardless of waiting method.

File: helpers/utils.js (hunk 4 above)
Comment (Yatskov) [reply helpers/utils.js]: I think we should handle this later.

**Pull request merged**